api/client/service.proto

```diff
@@ -34,10 +36,59 @@ message ExecutionResponse {
   google.rpc.Status error_detail = 7;
 }
 
+// Used to request a RequestStreamResponse.
+message RequestStreamRequest {
+  // Amount of RequestStreamResponses that should be yielded.
+  // Used to implement basic flow control by the client. Once AsyncClientImpl gets
+  // a way to apply backpressure this could probably be dropped.
+  uint64 amount = 1;
+}
+
+message RequestStreamResponse {
+  // Specifies what the requests look like. Will be merged with the client's configuration.
+  RequestSpecifier request_specifier = 1;
+  // Response-level expectations associated to the above request specification.
+  Expectations expectations = 2;
+}
+
+message RequestSpecifier {
+  // method/path/authority/content_length can all be expressed in request headers.
+  // so technically they are redundant. In case of ambiguity, the explicit fields will prevail.
+  // Note that as of today, the target upstream/cluster that will be used to replay traffic to
+  // fixes the destination address/port.
+  // Request method
+  google.protobuf.StringValue method = 1;
+  // Request path
+  google.protobuf.StringValue path = 2;
+  // Request authority
+  google.protobuf.StringValue authority = 3;
+  // Request content length. The client will transfer the number of bytes specified here for the
+  // request body.
+  google.protobuf.UInt32Value content_length = 4;
+  // Request header replacements. Any existing header(s) with the same name will be removed
+  // before setting.
+  envoy.api.v2.core.HeaderMap headers = 5;
+  // TODO(oschaaf): nice to have
+  // google.protobuf.StringValue sni_hostname = 10;
+}
+
+// Response-level expectations. If set, any unmet expectations will result in stats counter
+// increases in the future.
+message Expectations {
+  // Expected response status code
+  google.protobuf.UInt32Value response_code = 1;
```

Comment:
Is there already default behavior in NH around response_code validation?  In terms of requiring 200 responses or 2xx responses?  Are there possible scenarios where a consumer might want to specify not a specific response_code but a response_code range or array of possible response_codes?

I'm not positive that this design isn't sufficient, but just want to pose those questions just in case.

Reply:
Recently we landed the `termination-predicate` feature, which allows one to configure exit and failure conditions. It's a pretty basic implementation: you can state a counter threshold, and what happens when that gets exceeded: termination or failure. We atomically introduced default failure predicates to quickly stop upon observing connection failures and / or http error status codes. When one explicitly configures these predicates, that will override the default ones.
The idea was that this feature could also serve here; we track failed request-level expectations by bumping one or more counters, and then one could configure termination/failure-predicates to specify the exit condition and status.

Reply:
Thank you for the explanation!